### PR TITLE
Take care of all compiler warnings for GHC-9.8

### LINKED ISSUE #3990: Take care of all compiler warnings for GHC-9.8
Conditional in this snippet:
https://github.com/IntersectMBO/cardano-ledger/blob/92fd6688b3cce015abc46c133c68e6ea7baa6503/cabal.project#L69-L72

should be turned into ghc options that are always on:
```
program-options 
  ghc-options: -Werror
```

### PR DESCRIPTION
**Pull request by neilmayhew**
# Description

There are new warnings in ghc 9.8 for partial functions and orphan type family instances. This fixes the warnings and re-enables `-Werror` for all ghc versions.

Closes #3990

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by neilmayhew** [comment]
I don't really like enabling `-Wno-orphans` in larger modules that have a lot of other code in them, but the three affected modules are all in `cardano-ledger-test` so I didn't think it was worth trying to move the instances to a better place. However, I could look into this if it would be helpful.

File: libs/plutus-preprocessor/src/Debug.hs [hunk 1]
```diff
@@ -7,4 +7,4 @@ import Cardano.Ledger.Plutus.Evaluate (debugPlutus)
 import System.Environment (getArgs)
 
 main :: IO ()
-main = print . debugPlutus @StandardCrypto . head =<< getArgs
+main = mapM_ (print . debugPlutus @StandardCrypto) =<< getArgs
```
Comment (Lucsanszky) [inline libs/plutus-preprocessor/src/Debug.hs:10]: As discussed during the meeting: this should be moved from here in a separate PR and `opt-parse-applicative` should be used for more granularity.

**Review by Lucsanszky — APPROVED** [review]
LGTM! This should be good to go once it's rebased (and the new warnings are resolved). :+1:

File: libs/plutus-preprocessor/src/Debug.hs (hunk 1 above)
Comment (neilmayhew) [reply libs/plutus-preprocessor/src/Debug.hs:10]: I believe this already has an issue (#3920) so I'll resolve this conversation

**Review by lehins — APPROVED** [review]
Beautiful!
Thank you!
